_source/_posts/2022-07-21-angular-security-xss.md

```diff
@@ -0,0 +1,402 @@
+2. Style - binding styles to add more flair to the site
+3. URL - binding URLs like when you want to navigate to an external site in an anchor tag
+4. Resource URL - binding URLs that load and run as code
+
+To mark the value as trusted and safe to use, you can inject `DomSanitizer` and use one of the following methods appropriate for the security context to return a value marked as safe.
+1. `bypassSecurityHTML` 
+2. `bypassSecurityScript`
+3. `bypassSecurityTrustStyle` 
+4. `bypassSecurityTrustUrl` 
+5. `bypassSecurityTrustResourceUrl` 
+
+These methods return the same input but are marked as trusted by wrapping it in a safe equivalent of the sanitization type.
+
+Let's see what this component looks like when we mark the HTML value as trusted.
+
+```ts
+@Component({
+  selector: 'app-trustworthy-image',
+  template: `
+    <section [innerHTML]="html"
+  `
+})
+export class TrustworthyImageComponent {
+  public html = `<img src=1 onerror="alert('Doh!')"/>`;
+  public safeHtml: SafeHtml;
+
+  constructor(sanitizer: DomSanitizer) {
+    this.safeHtml = sanitizer.bypassSecurityTrustHtml(this.html);
+  }
+}
+```
+
+Now, if you view this in the browser, you'll see the broken image and an alert pop up. Success?? Maybe...
+
+![Giphy of not sure if meme](https://media.giphy.com/media/cFgb5p5e1My3K/giphy.gif)
+
+Let's look at an example with a resource URL, such as the DOM-based XSS example where we bind the URL of the `iframe` source. 
+
+Your component code might look like this
+```ts
+@Component({
+  selector: 'app-video',
+  template: `
+    <iframe [src]="linky" width="800px" height="450px"
+  `
+})
+export class VideoComponent {
+
+  // pretend this is from an external source
+  public linky = '//videolink/embed/12345';
+}
+```
+
+Angular will stop you right there. 🛑
+
+You'll see an error in the console saying unsafe values can't be used in a resource URL. Angular recognizes you're trying to add a resource URL and is alarmed you're doing something dangerous. Resource URLs can contain legitimate code, so Angular can't sanitize it, unlike the comments we had above.
+
+If we are **sure** our link is safe and trustworthy (highly debatable in this example, but we'll ignore that for a moment), we can mark the resource as trusted after we do some cleanup to make the resource URL safer.
+
+
+> Remember, we should only use the `bypassSecurityTrust...` methods if we know the value is trustworthy. You bypass the built-in security mechanisms and open yourself to security vulnerabilities!
+
+Instead of using the entire video URL based on the external party's API response, we'll construct the URL by defining the video host URL within our app and appending the video's ID that we get back from the external party's API response. This way, we aren't entirely reliant on a potentially untrustworthy value from a third party, and we'll have some measure of ensuring we're not injecting malicious code into the URL. 
```

Comment:
```suggestion
Instead of using the entire video URL based on the external party's API response, we'll construct the URL by defining the video host URL within our app and appending the video's ID that we get back from the external party's API response. This way, we aren't entirely reliant on a potentially untrustworthy value from a third party. Instead, we'll have some measure of ensuring we're not injecting malicious code into the URL. 
```
2 sentences for readability